compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -41,49 +35,8 @@ pub fn is_const_evaluatable<'cx, 'tcx>(
             if satisfied_from_param_env(tcx, ct, param_env)? {
                 return Ok(());
             }
-
-            // We were unable to unify the abstract constant with
-            // a constant found in the caller bounds, there are
-            // now three possible cases here.
-            #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
-            enum FailureKind {
-                /// The abstract const still references an inference
-                /// variable, in this case we return `TooGeneric`.
-                MentionsInfer,
-                /// The abstract const references a generic parameter,
-                /// this means that we emit an error here.
-                MentionsParam,
-                /// The substs are concrete enough that we can simply
-                /// try and evaluate the given constant.
-                Concrete,
-            }
-            let mut failure_kind = FailureKind::Concrete;
-            walk_abstract_const::<!, _>(tcx, ct, |node| match node.root(tcx) {
-                Node::Leaf(leaf) => {
-                    if leaf.has_infer_types_or_consts() {
-                        failure_kind = FailureKind::MentionsInfer;
-                    } else if leaf.has_param_types_or_consts() {
-                        failure_kind = cmp::min(failure_kind, FailureKind::MentionsParam);
-                    }
-
-                    ControlFlow::CONTINUE
-                }
-                Node::Cast(_, _, ty) => {
-                    if ty.has_infer_types_or_consts() {
-                        failure_kind = FailureKind::MentionsInfer;
-                    } else if ty.has_param_types_or_consts() {
-                        failure_kind = cmp::min(failure_kind, FailureKind::MentionsParam);
-                    }
-
-                    ControlFlow::CONTINUE
-                }
-                Node::Binop(_, _, _) | Node::UnaryOp(_, _) | Node::FunctionCall(_, _) => {
-                    ControlFlow::CONTINUE
-                }
-            });
-
-            match failure_kind {
-                FailureKind::MentionsInfer => {
+            match ct.unify_failure_kind(tcx) {
+                FailureKind::HasTyError | FailureKind::MentionsInfer => {
```

Comment:
with `HasTyError` / `Error` we should return `Ok(())` and not emit an additional error